include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,367 @@ enum E_DESCRIPTOR_TYPE : uint8_t
+public:
+	using sampler_type = SamplerType;
+
+	struct SBinding
 	{
+		enum class E_CREATE_FLAGS : uint8_t
+		{
+			ECF_NONE = 0,
+			ECF_UPDATE_AFTER_BIND_BIT = 1u << 1,
+			ECF_UPDATE_UNUSED_WHILE_PENDING_BIT = 1u << 2,
+			ECF_PARTIALLY_BOUND_BIT = 1u << 3
+		};
+
+		uint32_t binding;
+		IDescriptor::E_TYPE type;
+		core::bitflag<E_CREATE_FLAGS> createFlags;
+		core::bitflag<IShader::E_SHADER_STAGE> stageFlags;
+		uint32_t count;
+		// Use this if you want an immutable sampler that is baked into the DS layout itself.
+		// If its `nullptr` then the sampler used is mutable and can be specified while writing the image descriptor to a binding while updating the DS.
+		const core::smart_refctd_ptr<sampler_type>* samplers;
+	};
+
+	// Maps a binding to a local (to descriptor set layout) offset.
+	class CBindingRedirect
+	{
+	public:
 		static constexpr inline uint32_t Invalid = ~0u;
 
-		SBindingRedirect() : bindings(nullptr), offsets(nullptr), count(0ull) {}
+		struct binding_number_t
+		{
+			inline binding_number_t(const uint32_t d) : data(d) {}
+			uint32_t data;
+		};
+
+		struct storage_offset_t
+		{
+			inline storage_offset_t(const uint32_t d) : data(d) {}
+			uint32_t data;
+		};
+
+		inline uint32_t getBindingCount() const { return m_count; }
+
+		// Returns index into the binding property arrays below (including `m_storageOffsets`), for the given binding number `binding`.
+		// Assumes `m_bindingNumbers` is sorted and that there are no duplicate values in it.
+		inline uint32_t searchForBinding(const binding_number_t binding) const
+		{
+			if (!m_bindingNumbers)
+				return Invalid;
+
+			assert(m_storageOffsets && (m_count != 0u));
+
+			auto found = std::lower_bound(m_bindingNumbers, m_bindingNumbers + m_count, binding, [](binding_number_t a, binding_number_t b) -> bool {return a.data < b.data; });
+
+			if ((found >= m_bindingNumbers + m_count) || (found->data != binding.data))
+				return Invalid;
+
+			const uint32_t foundIndex = found - m_bindingNumbers;
+			assert(foundIndex < m_count);
+			return foundIndex;
+		}
+
+		inline binding_number_t getBindingNumber(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return m_bindingNumbers[index];
+		}
+
+		inline core::bitflag<IShader::E_SHADER_STAGE> getStageFlags(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return m_stageFlags[index];
+		}
 
-		SBindingRedirect(const size_t _count) : count(_count)
+		inline uint32_t getCount(const uint32_t index) const
 		{
-			bindings = std::make_unique<uint32_t[]>(count << 1);
-			offsets = bindings.get() + count;
+			assert(index < m_count);
+			return (index == 0u) ? m_storageOffsets[index].data : m_storageOffsets[index].data - m_storageOffsets[index - 1].data;
 		}
 
-		// For a given binding number `binding` in `bindings`, return its offset in `offsets`.
-		inline uint32_t operator[](const uint32_t binding) const
+		inline storage_offset_t getStorageOffset(const uint32_t index) const
```

Comment:
Use `storage_range_index_t` instead of the `uint32_t`, for more detail see https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/8fbc87079edc5109a9c916eeb7ea7c0eff9ce128..b3719ddb1377d8af8f8d5f4dd112d3c844394553#r1114576684

Reply:
Resolved.